src/plugins/workspace/public/plugin.ts

```diff
@@ -298,44 +293,11 @@ export class WorkspacePlugin
       core.http.basePath.getBasePath()
     );
 
-    if (workspaceId) {
-      const result = await workspaceClient.enterWorkspace(workspaceId);
-      if (!result.success) {
-        /**
-         * Fatal error service does not support customized actions
-         * So we have to use a self-hosted page to show the errors and redirect.
-         */
-        (async () => {
-          const [{ application, chrome }] = await core.getStartServices();
-          chrome.setIsVisible(false);
-          application.navigateToApp(WORKSPACE_FATAL_ERROR_APP_ID, {
-            replace: true,
-            state: {
-              error: result?.error,
-            },
-          });
-        })();
-      } else {
-        /**
-         * If the workspace id is valid and user is currently on workspace_fatal_error page,
-         * we should redirect user to overview page of workspace.
-         */
-        (async () => {
-          const [{ application }] = await core.getStartServices();
-          const currentAppIdSubscription = application.currentAppId$.subscribe((currentAppId) => {
-            if (currentAppId === WORKSPACE_FATAL_ERROR_APP_ID) {
-              application.navigateToApp(WORKSPACE_DETAIL_APP_ID);
-            }
-            currentAppIdSubscription.unsubscribe();
-          });
-          // Add workspace id to recent workspaces.
-          recentWorkspaceManager.addRecentWorkspace(workspaceId);
-        })();
-      }
-    }
+    this.workspaceValidationService.setup(core, workspaceId);
 
     const mountWorkspaceApp = async (params: AppMountParameters, renderApp: WorkspaceAppType) => {
       const [coreStart, { navigation }] = await core.getStartServices();
+      const workspaceClient = coreStart.workspaces.client$.getValue() as WorkspaceClient;
```

Comment:
I am thinking we can get the client from `this.workspaceValidationService.setup`.

Reply:
We will need to make `workspaceClient` public so I wonder is it appropriate to directly expose it  from `WorkspaceValidationService`?